### PR TITLE
Bugfix suggestions

### PR DESCRIPTION
**Pull request by carlosdelest**
Added a dedicated ActionListener class to deal with the common IndexNotFound onFailure management. LMKWYT!

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java [hunk 1]
```diff
@@ -488,5 +462,31 @@ static void writeSearchApplicationBinaryWithVersion(SearchApplication app, Outpu
         }
     }
 
+    static class DelegatingIndexNotFoundActionListener<T, R> extends ActionListener.Delegating<T, R> {
```
Comment (DaveCTurner) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:465]: Note that `ActionListener.Delegating` is now `org.elasticsearch.action.DelegatingActionListener` in `main` (#94320) so you'll get some conflict or compile error here after merging this and `main`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java [hunk 2]
```diff
@@ -488,5 +462,31 @@ static void writeSearchApplicationBinaryWithVersion(SearchApplication app, Outpu
...
+
+        @Override
+        public void onResponse(T t) {
+            bc.accept(delegate, t);
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (e instanceof IndexNotFoundException) {
+                delegate.onFailure(new ResourceNotFoundException(resourceName));
```
Comment (DaveCTurner) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java]: Might be useful to keep hold of the original stack trace here too - something like this should do the trick I think:

```suggestion
                delegate.onFailure(new ResourceNotFoundException(resourceName, e));
```

**Review by DaveCTurner — COMMENTED** [review]
It'd simplify things a little more to model this utility on `ActionListener#map` instead of `ActionListener#delegateFailure`. `delegateFailure` is a little more general, but I don't think you need the generality here.

**Comment by carlosdelest** [comment]
> It'd simplify things a little more to model this utility on ActionListener#map instead of ActionListener#delegateFailure. delegateFailure is a little more general, but I don't think you need the generality here.

@DaveCTurner , could you please provide an example? AFAIU map() would be useful for mapping the response, but it's the onFailure and the exception the ones we would want to map here....

Thanks for your comments BTW! 💯

**Review by kderusso — APPROVED** [review]
Nice optimizations, thanks!

**Pull request merged**